### PR TITLE
Very preliminary relaxed mode draft.

### PR DESCRIPTION
**Pull request by olonho**

### CODE REVIEW

File: runtime/src/main/cpp/Memory.cpp [hunk 1]
```diff
@@ -273,6 +279,14 @@ constexpr const char* MemoryStatistic::indexToName[];
 
 #endif  // COLLECT_STATISTIC
 
+inline bool isPermanentOrFrozen(ContainerHeader* container) {
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp]: Is it different from `isPermanentOrFrozen` declared in Memory.h?

File: runtime/src/main/cpp/MemoryPrivate.hpp [hunk 2]
```diff
@@ -21,8 +21,12 @@
 
 extern "C" {
 
+MODEL_VARIANTS(void, ReleaseHeapRef, const ObjHeader* object);
+
 void AddRefFromAssociatedObject(const ObjHeader* object) RUNTIME_NOTHROW;
-void ReleaseRefFromAssociatedObject(const ObjHeader* object) RUNTIME_NOTHROW;
+void ReleaseRefFromAssociatedObject(const ObjHeader* object);
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/MemoryPrivate.hpp]: Is `RUNTIME_NOTHROW` removed intentionally?

File: runtime/src/main/cpp/Memory.cpp [hunk 3]
```diff
@@ -1852,8 +1819,8 @@ OBJ_GETTER(readHeapRefLocked, ObjHeader** location, int32_t* spinlock) {
   if (container != nullptr)
     incrementRC<true>(container);
   unlock(spinlock);
-  if (container != nullptr)
-    enqueueDecrementRC</* CanCollect = */ true>(container);
+  if (value != nullptr)
+    ReleaseHeapRef(value);
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp]: Can this remove object before return? (like it was in `swapHeapRefLocked`).

File: runtime/src/main/cpp/Memory.cpp [hunk 4]
```diff
@@ -1825,12 +1792,12 @@ OBJ_GETTER(swapHeapRefLocked,
     shallRelease = oldValue != nullptr;
   }
   unlock(spinlock);
-  if (shallRelease) {
-    releaseHeapRef(oldValue);
-  }
   // No need to rememberNewContainer(), as oldValue is already
   // present on this worker.
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp]: Btw, is this correct? Doesn't seem so when `oldValue != expectedValue`.

File: runtime/src/main/cpp/Memory.cpp [hunk 5]
```diff
@@ -2393,38 +2482,55 @@ OBJ_GETTER(InitSharedInstanceStrict,
...
   zeroHeapRef(location);
 }
 
-void ZeroStackRef(ObjHeader** location) {
-  zeroStackRef(location);
+void ZeroStackRefStrict(ObjHeader** location) {
+  zeroStackRef<true>(location);
+}
+void ZeroStackRefRelaxed(ObjHeader** location) {
+  zeroStackRef<true>(location);
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp]: Is `true` correct here?

File: runtime/src/main/kotlin/kotlin/native/Platform.kt [hunk 6]
```diff
@@ -58,6 +66,19 @@ public object Platform {
...
+    /**
+     * Memory model binary was compiled with.
+     */
+    public val memoryModel: MemoryModel
+        get() = MemoryModel.values()[Platform_getMemoryModel()]
+
+    /**
+     * If binary was compiled in debug mode.
+     */
+    public val isDebugBinary: Boolean
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/kotlin/kotlin/native/Platform.kt]: Maybe related: #2562

File: runtime/src/main/cpp/Memory.cpp (hunk 4 above)
Comment (olonho) [reply runtime/src/main/cpp/Memory.cpp]: Comment was indeed wrong.

Comment (SvyatoslavScherbina) [reply runtime/src/main/cpp/Memory.cpp]: Comment is still wrong, and the behaviour is wrong too.
The following test crashes:
```kotlin
import kotlin.test.*
import kotlin.native.concurrent.*

val atomicRef = AtomicReference<Any?>(Any().freeze())
val semaphore = AtomicInt(0)

@Test fun runTest() {
    semaphore.value = 0
    val worker = Worker.start()
    val future = worker.execute(TransferMode.SAFE, { null }) {
        val value = atomicRef.compareAndSwap(null, null)
        semaphore.increment()
        while (semaphore.value != 2) {}
        println(value.toString() != "")
    }
    while (semaphore.value != 1) {}
    atomicRef.value = null
    kotlin.native.internal.GC.collect()
    semaphore.increment()
    future.result
    worker.requestTermination().result
}
```
(based on `worker10`).
Generally `atomicRef.compareAndSwap(Any().freeze(), null)` is the same as `atomicRef.value`, and if the latter requires `rememberNewContainer`, then the former requires this too.

Comment (olonho) [reply runtime/src/main/cpp/Memory.cpp]: Thanks for bringing my attention to this problem. Fixed.

File: backend.native/tests/runtime/workers/worker10.kt [hunk 7]
```diff
@@ -163,4 +163,22 @@ val stableHolder2 = StableRef.create(("hello" to "world").freeze())
     semaphore.increment()
     future.result
     worker.requestTermination().result
+}
+
+@Test fun runTest6() {
```
Comment (SvyatoslavScherbina) [inline backend.native/tests/runtime/workers/worker10.kt]: This test doesn't fail even without the fix applied.

File: runtime/src/main/cpp/Memory.cpp [hunk 8]
```diff
@@ -1794,10 +1794,13 @@ OBJ_GETTER(swapHeapRefLocked,
...
   UpdateReturnRef(OBJ_RESULT, oldValue);
-  // No need to rememberNewContainer(), as if `oldValue` is not null - it is explicitly released
-  // anyway, and thus can not escape GC.
   if (shallRelease) {
+    // No need to rememberNewContainer() on this path, as if `oldValue` is not null - it is explicitly released
+    // anyway, and thus can not escape GC.
     ReleaseHeapRef(oldValue);
+  } else {
+    if (IsStrictMemoryModel && oldValue != expectedValue)
+      rememberNewContainer(oldValue->container());
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Memory.cpp]: Isn't it too late to perform this after unlock?

File: backend.native/tests/runtime/workers/worker10.kt (hunk 7 above)
Comment (olonho) [reply backend.native/tests/runtime/workers/worker10.kt]: Yeah, better use fresh atomic ref here.

**Comment by omiwrench** [comment]
Could we get some documentation on how to enable this relaxed mode?

**Pull request merged**